generate_analytical_dataset.py

```diff
@@ -0,0 +1,274 @@
+    created_date_time = retrieve_date_time_element('createdDateTime', db_object)
+    if last_modified_date_time:
+        return last_modified_date_time
+    if created_date_time:
+        return created_date_time
+    return epoch
+
+def retrieve_date_time_element(key, db_object):
+    date_element = db_object.get(key)
+    if date_element is not None:
+        if isinstance(date_element, dict):
+            date_sub_element =  date_element.get('$date')
+            if date_sub_element is not None:
+                return date_sub_element
+        else:
+            return date_element
+    return ""
+
+def format_date_to_valid_outgoing_format(current_date_time):
+    parsed_date_time = get_valid_parsed_date_time(current_date_time)
+    parsed_date_time.astimezone(pytz.utc)
+    parsed_date_time = datetime.strftime(parsed_date_time, '%Y-%m-%dT%H:%M:%S.%f')
+    (dt, micro) = parsed_date_time.split('.')
+    dt = "%s.%03d%s" % (dt, int(micro) / 1000, 'Z')
+    return dt
+
+def get_valid_parsed_date_time(time_stamp_as_string):
+    valid_timestamps =  ['%Y-%m-%dT%H:%M:%S.%fZ','%Y-%m-%dT%H:%M:%S.%f%z']
+    for time_stamp_fmt in valid_timestamps:
+        try:
+            return datetime.strptime(time_stamp_as_string, time_stamp_fmt)
+        except Exception:
+            print(f'timestampAsString did not match valid format {time_stamp_fmt}')
+
+    # TODO Think about the below exception later on
+    raise Exception(f'Unparseable date found: {time_stamp_as_string} , did not match any supported date formats {valid_timestamps}')
+
+def sanitize(z):
+    decrypted = z['decrypted']
+    db_name = z['db_name']
+    collection_name = z['collection_name']
+    if ((db_name == "penalties-and-deductions" and collection_name == "sanction")
+            or (db_name == "core" and collection_name == "healthAndDisabilityDeclaration")
+            or (db_name == "accepted-data" and collection_name == "healthAndDisabilityCircumstances")):
+        decrypted = re.sub(r'(?<!\\)\\[r|n]', '', decrypted)
+    return decrypted.replace("$", "d_").replace("\u0000", "").replace("_archivedDateTime", "_removedDateTime").replace("_archived", "_removed")
+
+def decrypt(y):
+    key = y['plain_text_key']
+    iv = y['iv']
+    ciphertext = y['dbObject']
+    iv_int = int(binascii.hexlify(base64.b64decode(iv)), 16)
+    ctr = Counter.new(AES.block_size * 8, initial_value=iv_int)
+    aes = AES.new(base64.b64decode(key), AES.MODE_CTR, counter=ctr)
+    y['decrypted'] = aes.decrypt(base64.b64decode(ciphertext))
+    return y
+
+def get_plaintext_key_calling_dks(r,keys_map):
+    cek = r['encryptedKey']
+    kek = r['keyEncryptionkeyId']
+    if keys_map.get(cek):
+        key = keys_map[cek]
+        print("Found the key in cache")
+        r['plain_text_key'] = key
+    else:
+        print("Didn't find the key in cache so calling dks")
+        key = call_dks(cek, kek)
+        keys_map[cek] = key
+        r['plain_text_key'] = key
+    return r
```

Comment:
I am thinking that after some time it will be difficult to remember what parameters p, z, y and r are

Reply:
Thanks Abdu . Changed it to a meaningful name now